Patch out insecure dependencies and add NixOS instructions

IvarWithoutBones opened the pull request:
This commit uses `patchelf` to remove the dependency on OpenSSL 1.0, which is outdated and considered insecure. It replaces this dependency with libcurl, of which the latest version works just fine. I tested the multiplayer mod as well and didn't notice any issues.

This should make sure that any distrobution can run the generated binary out of the box, without needing a dynamically linked placeholder.

I also noticed the game failed to start on my AMD machine, and after looking through Mesa's Gitlab I found a workaround using an environment variable. A wrapper bash script that sets this variable and executes the `runner` binary is added to take care of this automatically.

This commit adds instructions on how to run the game on NixOS as well. I will also PR a derivation to nixpkgs (our package repository) to make the process of running the game a bit more seamless in the future.

Miepee commented on README.md:
```diff
@@ -2,7 +2,7 @@
 
 ## Dependencies
-The patcher and the installer only require a small amount of dependencies:
+The patcher and the installer only require a small amount of dependencies. Note that the `java`, `jdk` or `jre` packages are only required for building an Android APK.
```
Maybe rewrite this as `Note that the Java-related packages`?

IvarWithoutBones replied:
I've added that, but still included the potential names of said packages. Its possible for users not to know all of these are java related.

Miepee commented on README.md:
```diff
@@ -13,28 +13,34 @@ Include = /etc/pacman.d/mirrorlist
+Enter a shell with all the required dependencies to run the script:
+`nix shell nixpkgs#{gnused,unzip,xdelta,patchelf,jre}`
+
+After running the script the game can be started using the `steam-run` FHS environment, as to avoid having to patch its dynamic linker and dynamically linked libraries. You can run it with the following command:
```
Not that familiar with Nix, what does FHS stand for? Is it a common abbreviation?

IvarWithoutBones replied:
Its an environment with the normal path to the dynamic linker existing, which it does not on NixOS. Paths like `/usr/lib` are hardcoded in the binary but not present. Tools like `steam-run` provide an environment where they are present,

Miepee commented on README.md:
```diff
@@ -13,28 +13,34 @@ Include = /etc/pacman.d/mirrorlist
+Enter a shell with all the required dependencies to run the script:
+`nix shell nixpkgs#{gnused,unzip,xdelta,patchelf,jre}`
+
+After running the script the game can be started using the `steam-run` FHS environment, as to avoid having to patch its dynamic linker and dynamically linked libraries. You can run it with the following command:
```
```diff
- After running the script the game can be started
+ After running the script, the game can be started
```

Miepee commented on README.md:
```diff
@@ -13,28 +13,34 @@ Include = /etc/pacman.d/mirrorlist
+`nix shell nixpkgs#{gnused,unzip,xdelta,patchelf,jre}`
+
+After running the script the game can be started using the `steam-run` FHS environment, as to avoid having to patch its dynamic linker and dynamically linked libraries. You can run it with the following command:
+`NIXPKGS_ALLOW_UNFREE=1 nix shell --impure nixpkgs#steam-run --command steam-run ./runner`
```
Command would not work for the Appimage. A small notice would probably be good.
Also, since the section at the top if for running the generated AppImage, are there any differences between running that, and running the "pure" binary? If yes, then those should probably be also clarified at the bottom under the "manual installation" section.

IvarWithoutBones replied:
> Also, since the section at the top if for running the generated AppImage, are there any differences between running that, and running the "pure" binary

The appimage does not work on NixOS due to outdated/insecure dependencies. We prefer not to use them anways. I put in a notice.

> If yes, then those should probably be also clarified at the bottom under the "manual installation" section.

Done :+1:

Miepee commented on README.md:
```diff
@@ -69,31 +75,18 @@ If for some reason, you are not able to access your phone, you can use the adb (
-
-On newer versions (Debian 10+ or Ubuntu 18+) you may have to do the below command instead. Please make sure first, that you neither have `libcurl3` nor `libcurl4` installed.  
-`sudo apt install unzip sed xdelta3 libc6:i386 libstdc++6:i386 zlib1g-dev:i386 libxxf86vm1:i386 libopenal1:i386 libxrandr2:i386 libglu1:i386 && wget http://archive.ubuntu.com/ubuntu/pool/main/o/openssl1.0/libssl1.0.0_1.0.2n-1ubuntu5.5_i386.deb && sudo dpkg -i libssl1.0.0_1.0.2n-1ubuntu5.5_i386.deb && wget http://archive.ubuntu.com/ubuntu/pool/universe/c/curl3/libcurl3_7.58.0-2ubuntu2_i386.deb &&  ...
+`sudo apt install unzip sed xdelta3 libc6:i386 libstdc++6:i386 zlib1g-dev:i386 libxxf86vm1:i386 libcurl:i386 libopenal1:i386 libxrandr2:i386 libglu1:i386`
```
I just realized, a mention to enable multiarch is missing on here.

Miepee commented on README.md:
```diff
@@ -69,31 +75,18 @@ If for some reason, you are not able to access your phone, you can use the adb (
 ## Manual installation
 This section, is if you don't like appimages, or want to run the game natively. Unless you have a very good reason for it, please use the AppImage instead!
```
Just realized, can you properly capitalize AppImage here too please?

Miepee replied:
Seems like this was missed, whoops.
For reference: 
```diff
- This section, is if you don't like appimages
+ This section, is if you don't like AppImages
```

Miepee commented on README.md:
````diff
@@ -69,31 +75,18 @@ If for some reason, you are not able to access your phone, you can use the adb (
-gcc -fPIC -shared -Wl,-soname,libcurl.so.4 empty.o -o libcurl.so.4
-```
-And then launch AM2R with LD_LIBRARY_PATH pointing to where your generated .so file is. For example `LD_LIBRARY_PATH=/home/me/libcurl-lib ./AM2R`
+`sudo yum install unzip sed xdelta openal-soft openssl`
````
Wouldn't all of these still need patchelf?

IvarWithoutBones replied:
Yes.

Miepee commented on patcher.sh:
```diff
@@ -66,6 +66,12 @@ patch_am2r ()
 		fi
 
+		# Check whether patchelf is installed
+		if [ !  -x "$(command -v patchelf)" ] ; then
```
At this point, might as well create a function for it, rather than copy-pasting the same thing 4 times or so.

Miepee commented on patcher.sh:
```diff
@@ -66,6 +66,12 @@ patch_am2r ()
 
+		# Check whether patchelf is installed
+		if [ !  -x "$(command -v patchelf)" ] ; then
+			>&2 echo "patchelf is not installed! Please install 'patchelf' from your local package manager!"
```
```diff
- patchelf is not installed!
+ Patchelf is not installed!
```

Miepee commented on patcher.sh:
```diff
@@ -101,16 +107,38 @@ patch_am2r ()
-			fi
-		fi
+                # GameMaker games (like AMR2) link to OpenSSL 1.0.0, which is outdated and insecure.
+                # When attempting to link to newer versions an error is raised at runtime claiming it cannot find
```
```suggestion
                # When attempting to link to newer versions, an error is raised at runtime claiming it cannot find
```

Miepee commented on patcher.sh:
```diff
@@ -101,16 +107,38 @@ patch_am2r ()
+                # GameMaker games (like AMR2) link to OpenSSL 1.0.0, which is outdated and insecure.
+                # When attempting to link to newer versions an error is raised at runtime claiming it cannot find
+                # the outdated version of OpenSSL, even though it has been patched to link to the newer version.
+                # After replacing it with libcurl versioning is ignored, and the binary starts just fine.
```
```suggestion
                # After replacing it with libcurl, versioning is ignored, and the binary starts just fine.
```

Miepee commented on patcher.sh:
```diff
@@ -101,16 +107,38 @@ patch_am2r ()
+                # When attempting to link to newer versions an error is raised at runtime claiming it cannot find
+                # the outdated version of OpenSSL, even though it has been patched to link to the newer version.
+                # After replacing it with libcurl versioning is ignored, and the binary starts just fine.
+                echo "Replacing insecure OpenSSL dependency with libcurl..."
```
Personally prefer "Patching" rather than "Replacing"

Miepee commented on patcher.sh:
```diff
@@ -101,16 +107,38 @@ patch_am2r ()
+                        --replace-needed "libssl.so.1.0.0" "libcurl.so"
+
+                # An environment variable needs to be set on Mesa to avoid a race related to multithreaded shader compilation.
+                # To do this we move the original executable to a hidden file, and create a bash script with the needed variable in place of the original.
```
```diff
- # To do this we move the original executable to a hidden file, and create a bash script with the needed variable in place of the original.
+ # To do this, we move the original executable to a hidden file, and create a bash script with the needed variable in place of the original.

Miepee commented on patcher.sh:
```diff
@@ -101,16 +107,38 @@ patch_am2r ()
+                        radeonsi_sync_compile="true" exec "$(dirname "${BASH_SOURCE[0]}")/.runner-unwrapped" "$@"
+                ' | sed -e 's/^[[:space:]]*//' \
+                  | sed -r '/^\s*$/d' > "$GAMEDIR/runner"
+                # Note that the sed commands above remove leading whitespace and empty lines.
```
Personally, instead of using sed to fix the output afterwards, I prefer to just use this for cases like these:
```bash
                echo '
#!/usr/bin/env bash
# This environment variable fixes Mesa support. If another driver is used this should not do anything.
# See https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/4181 for more information.
radeonsi_sync_compile="true" exec "$(dirname "${BASH_SOURCE[0]}")/.runner-unwrapped" "$@"
                '
```

Miepee commented on patcher.sh:
```diff
@@ -101,16 +107,38 @@ patch_am2r ()
+                        else
+                                ln -sf "$GAMEDIR/runner" "$PREFIX/bin/am2r"
+                        fi
+                fi
```
(Lowest line I can comment on, imagine this is the AppImage creation line)
In the current state, the AppImage still ships the outdated crypto/ssl libs. Do you want to remove those (in the patching process, the AM2RLauncher would break if you'd remove them from the repo)? AppImages aren't really sandboxed, so there may be security implications by them.

IvarWithoutBones replied:
> In the current state, the AppImage still ships the outdated crypto/ssl libs. Do you want to remove those (in the patching process, the AM2RLauncher would break if you'd remove them from the repo)? AppImages aren't really sandboxed, so there may be security implications by them.

I cannot test appimage creation myself, and I dont see where the insecure dependencies are included. I do see a bunch of `so`s being included in this repository, but ones compiled on NixOS would not work well in that environment. I would appreciate this if someone else could do that in a followup PR.

Miepee replied:
Links to [libssl-100](https://github.com/AM2R-Community-Developers/AM2R-Autopatcher-Linux/blob/master/data/AM2R.AppDir/usr/lib32/libssl.so.1.0.0) and [libcrypto-100](https://github.com/AM2R-Community-Developers/AM2R-Autopatcher-Linux/blob/master/data/AM2R.AppDir/usr/lib32/libcrypto.so.1.0.0)
The whole AM2R.AppDir folder gets copied / used as a base for the AppImage.

> I would appreciate this if someone else could do that in a followup PR

Would also be fine for me.

Miepee reviewed (commented):
Some small notes
Also, this should require testing on Intel and Nvidia GPUs as well, since last time I checked, those also had specific environment variables to get around those issues. Don't have any specific PRs to link, but see the FAQ question for [Intel here](https://am2r-community-developers.github.io/DistributionCenter/faq.html#i-am-on-ubuntu-have-an-intel-gpu-and-the-game-doesnt-launch) and for [Nvidia here](https://am2r-community-developers.github.io/DistributionCenter/faq.html#i-have-an-nvidia-gpu-with-proprietary-drivers-and-the-game-doesnt-launch)

IvarWithoutBones committed:
Patch out insecure dependencies and add NixOS instructions

This commit uses `patchelf` to remove the dependency on OpenSSL 1.0,
which is outdated and considered insecure. It replaces this dependency
with libcurl, of which the latest version works just fine.

This should make sure that any distrobution can run the generated binary
out of the box, without needing a dynamically linked placeholder.

I also noticed the game failed to start on my AMD machine, and after looking
through Mesa's Gitlab I found a workaround using an environment variable.
A wrapper bash script that sets this variable and executes the `runner` binary
is created to take care of this automatically.

This commit adds instructions on how to run the game on NixOS as well.
I will also PR it to nixpkgs (our package repository) to make the
process a bit more seamless in the future.

IvarWithoutBones commented:
Rebased this PR and addressed feedback :+1:

Miepee commented on patcher.sh:
```diff
@@ -19,8 +19,20 @@ VERSION="1.5.5"
 
+# Check wether a command is installed
+checkInstalled() {
+        local command="$1"
```
Formatting seems weird here.

IvarWithoutBones replied:
What would you prefer?

Miepee replied:
![image](https://user-images.githubusercontent.com/38186597/184553553-5e17ca2a-3d01-4538-abb7-c8348bf268a5.png)
the variable assignment and comment should be aligned with the if condition.

IvarWithoutBones replied:
Should be good now :+1: 

Edit: Discussion in DMs has shown this is a bit inconsistent, looks fine on my end but not on theirs. Will be fixed in a later commit adding an editorconfig file.

IvarWithoutBones commented:
> Also, this should require testing on Intel and Nvidia GPUs as well, since last time I checked, those also had specific environment variables to get around those issues. Don't have any specific PRs to link, but see the FAQ question for [Intel here](https://am2r-community-developers.github.io/DistributionCenter/faq.html#i-am-on-ubuntu-have-an-intel-gpu-and-the-game-doesnt-launch) and for [Nvidia here](https://am2r-community-developers.github.io/DistributionCenter/faq.html#i-have-an-nvidia-gpu-with-proprietary-drivers-and-the-game-doesnt-launch)

Works perfectly on an NVidia GPU. I dont have access to one from intel, but since they're using mesa as well im fairly certaing the fix works there as well.

Miepee commented on patcher.sh:
```diff
@@ -59,12 +71,8 @@ patch_am2r ()
-			>&2 echo "Xdelta is not installed! Please install 'xdelta3' from your local package manager!"
-			exit 1
-		fi
+                checkInstalled "xdelta3"
```
Considering xdelta is used in both Linux and Android, might also move it out to below the sed check.

IvarWithoutBones replied:
Sure :+1:

Miepee commented:
> The flags linked on said FAQ should not be required at all

Yup, we talked about this in DMs after, the FAQ-env vars were needed back then when we shipped the outdated GPU libraries.

Lojemiru commented on README.md:
```diff
@@ -2,39 +2,45 @@
 
 ### Debian (including Ubuntu, Mint, PopOS, etc.)
-Make sure that the i386 architecture is enabled for you, as libopenal1 requires the 32-bit version, and Ubuntu does not enable 32-bit support by default.  
+Make sure that the i386 architecture is enabled for you, as libopenal1 requires the 32-bit version, and Ubuntu does not enable 32-bit support by default.
```
Should be "Make sure you enable the i386 architecture, ..."

Remove comma after 'version'.

Lojemiru commented on README.md:
```diff
@@ -2,39 +2,45 @@
 
 ### Arch (including Manjaro, EndeavourOS, RebornOS, etc.)
-Make sure that multilib is enabled, as `lib32-libpulse` is a 32-bit library, and Arch does not enable 32-bit support by default.  
+Make sure that multilib is enabled, as `lib32-libpulse` is a 32-bit library, and Arch does not enable 32-bit support by default.
```
Remove comma after 'library'.

Miepee reviewed (approved):
Looks good for me.

Lojemiru commented on README.md:
```diff
@@ -2,39 +2,45 @@
+Enter a shell with all the required dependencies to run the script:
+`nix shell nixpkgs#{gnused,unzip,xdelta,patchelf,jre}`
+
+Note that the AppImage will **not** work on NixOS, please follow the instructions in the manual installation section.
```
Should be a period or semicolon instead of comma.

Lojemiru commented on README.md:
```diff
@@ -2,39 +2,45 @@
 
 ## Controllers
-In order to be able to use a controller, you *need* the two following packages as well:
+On some distributions, in order to be able to use a controller, you *need* the two following packages as well:
```
Should be "In order to use a controller on some distributions, the two following packages are *mandatory*:"

Lojemiru commented on README.md:
```diff
@@ -67,33 +73,29 @@ If for some reason, you are not able to access your phone, you can use the adb (
 
 ## Manual installation
-This section, is if you don't like appimages, or want to run the game natively. Unless you have a very good reason for it, please use the AppImage instead!
+This section, is if you don't like AppImages, or want to run the game natively. Unless you have a very good reason for it, please use the AppImage instead!
```
Should be "Manual installation should not be used without due cause. Unless you have a very good reason for avoiding AppImages, please use the default installation instead!"

Lojemiru commented on README.md:
```diff
@@ -67,33 +73,29 @@ If for some reason, you are not able to access your phone, you can use the adb (
 
 ### Debian (including Ubuntu, Mint, PopOS, etc.)
-`sudo apt install unzip sed xdelta3 libc6:i386 libstdc++6:i386 zlib1g-dev:i386 libxxf86vm1:i386 libcurl3:i386 libssl1.0:i386 libopenal1:i386 libxrandr2:i386 libglu1:i386`
+Make sure that the i386 architecture is enabled for you, as libopenal1 requires the 32-bit version, and Ubuntu does not enable 32-bit support by default.
```
Should be "Make sure you enable the i386 architecture, ..."

Remove comma after 'version'.

Lojemiru commented on README.md:
```diff
@@ -67,33 +73,29 @@ If for some reason, you are not able to access your phone, you can use the adb (
-On newer versions (Debian 10+ or Ubuntu 18+) you may have to do the below command instead. Please make sure first, that you neither have `libcurl3` nor `libcurl4` installed.  
-`sudo apt install unzip sed xdelta3 libc6:i386 libstdc++6:i386 zlib1g-dev:i386 libxxf86vm1:i386 libopenal1:i386 libxrandr2:i386 libglu1:i386 && wget http://archive.ubuntu.com/ubuntu/pool/main/o/openssl1.0/libssl1.0.0_1.0.2n-1ubuntu5.5_i386.deb && sudo dpkg -i libssl1.0.0_1.0.2n-1ubuntu5.5_i386.deb && wget http://archive.ubuntu.com/ubuntu/pool/universe/c/curl3/libcurl3_7.58.0-2ubuntu2_i386.deb &&  ...
+### Nix (Including NixOS)
+After running the script, the game can be started using the `steam-run` FHS environment, as to avoid having to patch its dynamic linker and dynamically linked libraries. You can run it with the following command:
```
Should be "After installation, the game can be run using the `steam-run` FHS environment. This avoids having to patch the dynamic linker and dynamically linked libraries. You can run it with the following command:"

Lojemiru reviewed (changes requested):
Content is good, grammar's a bit less so lol. Should be clear to merge after resolving these.

IvarWithoutBones committed:
readme: gramar fixes

IvarWithoutBones commented:
> Content is good, grammar's a bit less so lol. Should be clear to merge after resolving these.

Pushed fixes to all your suggestions. Most of these comments were an issue before my PR, but we might as well get them fixed now :)

Btw, when you have something you would like to suggest in CR, you can use suggestion blocks as follows:
![image](https://user-images.githubusercontent.com/41924494/184554587-de94afea-2e89-4f24-a46d-c0464036a1d0.png)

It'll properly format your suggestion which makes it a lot easier to read. Note that the contents dont really matter here, its just an example from a previous CR I did.

IvarWithoutBones committed:
patcher: fix syntax error

Miepee commented:
Also quickly chiming in: no need to type ` ```suggestion `, GH has an icon next to the other markdown elements that inserts one automatically.

IvarWithoutBones committed:
readme: add some codeblocks

Miepee commented:
Seems good, thanks for the PR! :)
